File: aat/config/enums.py
from enum import Enum


class BaseEnum(Enum):
    def __str__(self):
        return f'{self.value}'


class TradingType(BaseEnum):
    LIVE = 'LIVE'
    SIMULATION = 'SIMULATION'
    SANDBOX = 'SANDBOX'
    BACKTEST = 'BACKTEST'


class Side(BaseEnum):
    BUY = 'BUY'
    SELL = 'SELL'


class OptionType(BaseEnum):
    CALL = 'CALL'
    PUT = 'PUT'


class EventType(BaseEnum):
    # Heartbeat events
    HEARTBEAT = 'HEARTBEAT'

    # Trade events
    TRADE = 'TRADE'

    # Order events
    OPEN = 'OPEN'
    CANCEL = 'CANCEL'
    CHANGE = 'CHANGE'
    FILL = 'FILL'

    # Other data events
    DATA = 'DATA'

    # System events
    HALT = 'HALT'
    CONTINUE = 'CONTINUE'

    # Engine events
    ERROR = 'ERROR'
    START = 'START'
    EXIT = 'EXIT'

    # Order Events
    BOUGHT = 'BOUGHT'
    SOLD = 'SOLD'
    REJECTED = 'REJECTED'
    CANCELED = 'CANCELED'


class DataType(BaseEnum):
    DATA = 'DATA'
    ERROR = 'ERROR'

    ORDER = 'ORDER'
    TRADE = 'TRADE'


class InstrumentType(BaseEnum):
    OTHER = 'OTHER'

    EQUITY = 'EQUITY'

    # TODO ETF separate?

    BOND = 'BOND'

    OPTION = 'OPTION'

    FUTURE = 'FUTURE'

    PAIR = 'PAIR'

    SPREAD = 'SPREAD'

    FUTURESOPTION = 'FUTURESOPTION'

    MUTUALFUND = 'MUTUALFUND'

    COMMODITIES = 'COMMODITIES'

    # TODO Warrant?

    # Non-tradeable
    CURRENCY = 'CURRENCY'
    INDEX = 'INDEX'


class OrderType(BaseEnum):
    # Order Types
    LIMIT = 'LIMIT'
    MARKET = 'MARKET'
    STOP = 'STOP'


class OrderFlag(BaseEnum):
    # Order Flag
    NONE = 'NONE'
    FILL_OR_KILL = 'FILL_OR_KILL'
    ALL_OR_NONE = 'ALL_OR_NONE'
    IMMEDIATE_OR_CANCEL = 'IMMEDIATE_OR_CANCEL'

File: rust/aat-core/src/core/enums.rs
use serde::{Deserialize, Serialize};

macro_rules! all_enum_derivs{
    (
        $vis:vis enum $enum_name:ident {
            $(
                $field_vis:vis $field_name:ident
            ),*$(,)+
        }
    ) => {
        #[allow(non_camel_case_types)]
        #[derive(Copy, Clone, Debug, Serialize, Deserialize, PartialEq)]
        pub enum $enum_name{
            $($field_name,)*
        }
    }
}

all_enum_derivs! {
    pub enum TradingType {
        LIVE,
        SIMULATION,
        SANDBOX,
        BACKTEST,
    }
}

all_enum_derivs! {
    pub enum Side {
        BUY,
        SELL,
    }
}

all_enum_derivs! {
    pub enum OptionType {
        CALL,
        PUT,
    }
}

all_enum_derivs! {
    pub enum EventType {
        // Heartbeat events
        HEARTBEAT,

        // Trade events
        TRADE,

        // Order events
        OPEN,
        CANCEL,
        CHANGE,
        FILL,

        // Other data events
        DATA,

        // System events
        HALT,
        CONTINUE,

        // Engine events
        ERROR,
        START,
        EXIT,

        // Order Events
        BOUGHT,
        SOLD,
        RECEIVED,
        REJECTED,
        CANCELED,
    }
}

all_enum_derivs! {
    pub enum InstrumentType {
        OTHER,
        EQUITY,
        BOND,
        OPTION,
        FUTURE,
        SPREAD,
        FUTURESOPTION,
        PERPETUALFUTURE,
        MUTUALFUND,
        COMMODITY,
        CURRENCY,
        PAIR,
        INDEX,
    }
}

impl Default for InstrumentType {
    fn default() -> Self {
        InstrumentType::EQUITY
    }
}

all_enum_derivs! {
    pub enum OrderType {
        LIMIT,
        MARKET,
        STOP,
    }
}

all_enum_derivs! {
    pub enum OrderFlag {
        NONE,
        FILL_OR_KILL,
        ALL_OR_NONE,
        IMMEDIATE_OR_CANCEL,
    }
}

impl Default for OrderFlag {
    fn default() -> Self {
        OrderFlag::NONE
    }
}

all_enum_derivs! {
    pub enum ExitRoutine {
        NONE,
        CLOSE_ALL,
    }
}
